### PR TITLE
Replace RepoRef by Repo trait

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: lib/tests/test_revset.rs [hunk 1]
```diff
@@ -222,78 +222,78 @@ fn test_resolve_symbol_change_id(readonly: bool) {
         "040031cb4ad0cbc3287914f1d205dabf4a7eb889"
     );
 
-    let repo;
-    let repo_ref = if readonly {
-        repo = tx.commit();
-        repo.as_repo_ref()
+    let mut _readonly_repo = None;
+    let repo: &dyn Repo = if readonly {
+        _readonly_repo = Some(tx.commit());
```
Comment (yuja) [inline lib/tests/test_revset.rs:227]: `Option` isn't needed here. `_readonly_repo = tx.commit();` should be enough to keep it alive.

File: lib/tests/test_revset.rs [hunk 2]
```diff
@@ -531,12 +531,7 @@ fn test_evaluate_expression_root_and_checkout(use_git: bool) {
         .set_wc_commit(WorkspaceId::default(), commit1.id().clone())
         .unwrap();
     assert_eq!(
-        resolve_commit_ids_in_workspace(
-            mut_repo.as_repo_ref(),
-            "@",
-            &test_workspace.workspace,
-            None,
-        ),
+        resolve_commit_ids_in_workspace(mut_repo, "@", &test_workspace.workspace, None,),
```
Comment (yuja) [inline lib/tests/test_revset.rs:534]: nit: excessive comma

```
% grep -n ',)' lib/tests/test_revset.rs
534:        resolve_commit_ids_in_workspace(mut_repo, "@", &test_workspace.workspace, None,),
676:        resolve_commit_ids_in_workspace(mut_repo, "@-", &test_workspace.workspace, None,),
830:        resolve_commit_ids(mut_repo, &format!(":({}-)", commit4.id().hex()),),
```

File: lib/src/repo.rs [hunk 3]
```diff
@@ -625,27 +658,10 @@ impl MutableRepo {
         RepoRef::Mutable(self)
     }
 
-    pub fn base_repo(&self) -> &Arc<ReadonlyRepo> {
+    pub fn readonly_repo(&self) -> &Arc<ReadonlyRepo> {
```
Comment (yuja) [inline lib/src/repo.rs:581]: Just a comment. We could instead implement `Repo` for `Arc<ReadonlyRepo>`, and require `base_repo() -> &Arc<...>`. However, most callers don't need `&Arc`, so it wouldn't be worth.

**Review by yuja — APPROVED** [review]
Nice.

File: lib/tests/test_revset.rs (hunk 1 above)
Comment (martinvonz) [reply lib/tests/test_revset.rs]: Ah, good point. The variable is never used, so it doesn't need to be initialized in all code paths.

File: lib/src/repo.rs (hunk 3 above)
Comment (martinvonz) [reply lib/src/repo.rs]: Good point. I had not even considered that. Yes, most callers don't need `&Arc`, but almost all of them already have it, so maybe it's worth it after all? I'll add a commit on top to show what it looks like. I like. If you agree, I'll clean up the series.

Comment (yuja) [reply lib/src/repo.rs]: Indeed, `Arc<ReadonlyRepo>` is the data type we use everywhere, so new version looks good to me.

Comment (martinvonz) [reply lib/src/repo.rs]: I squashed in most of that last commit, but also added a tiny new commit early in the series. Feel free to take another look and mark this resolved (I'll enable auto-merge).

Comment (yuja) [reply lib/src/repo.rs]: LGTM, thanks.

**Pull request merged**